### PR TITLE
Enable external mode using SSL with RGW

### PR DESCRIPTION
**Pull request by dahorak**
Related to: https://url.corp.redhat.com/5479bff
Jenkins jobs related changes: https://url.corp.redhat.com/7c604db

### CODE REVIEW

**Comment by dahorak** [comment]
Verificaiton jobs
* without SSL for RGW https://url.corp.redhat.com/769755b (running)
* with SSL for RGW https://url.corp.redhat.com/03dd5c6 - the job failed because of RHBZ#2153199, but Proxy object was correctly updated with the CA certificate used for RGW with SSL:
```
08:28:17 - MainThread - ocs_ci.ocs.ocp - INFO  - Check if resource: ocs-ca-bundle exists.
08:28:17 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc --kubeconfig /home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig -n openshift-config get ConfigMap ocs-ca-bundle -n openshift-config -o yaml
08:28:18 - MainThread - ocs_ci.ocs.ocp - INFO  - Resource: ocs-ca-bundle, selector: None found.
08:28:18 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc --kubeconfig /home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig -n openshift-config get ConfigMap ocs-ca-bundle -n openshift-config -o yaml
08:28:18 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc -n openshift-config delete ConfigMap ocs-ca-bundle
08:28:18 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc create configmap ocs-ca-bundle -n openshift-config --from-file=ca-bundle.crt=/tmp/rgw-cert-ca0ze9bjl6.pem
08:28:18 - MainThread - ocs_ci.utility.utils - INFO  - Executing command: oc patch proxy/cluster --type=merge --patch='{"spec":{"trustedCA":{"name":"ocs-ca-bundle"}}}'
08:28:18 - MainThread - ocs_ci.utility.utils - INFO  - Sleeping for 60 sec to start update machineconfigpool status
08:29:18 - MainThread - ocs_ci.utility.utils - INFO  - Checking machineconfigpool status for worker nodes
```

and also the external cluster exporter script was correctly triggered with proper rgw-endpoint IP and port and --rgw-tls-cert-path:

```
08:42:51 - MainThread - ocs_ci.utility.connection - INFO  - Executing cmd: python3 /tmp/external-cluster-details-exporter-xw0xhbtv.py --rbd-data-pool-name rbd --rgw-endpoint <IP1>:443 --rgw-tls-cert-path /tmp/rgw-cert-ca.pem on <IP2>
```

File: ocs_ci/deployment/helpers/external_cluster_helpers.py [hunk 1]
```diff
@@ -396,6 +455,30 @@ def generate_exporter_script():
     return external_cluster_details_exporter.name
 
 
+def get_and_apply_rgw_cert_ca():
+    """
+    Downloads CA Certificate of RGW if SSL is used and apply it to be trusted
+    by the OCP cluster
+
+    Returns:
+        str: path to the downloaded RGW Cert CA
```
Comment (vavuthu) [inline ocs_ci/deployment/helpers/external_cluster_helpers.py:464]: new line below

File: ocs_ci/utility/ssl_certs.py [hunk 2]
```diff
@@ -273,6 +264,40 @@ def configure_custom_ingress_cert():
     wait_for_machineconfigpool_status("all", timeout=1800)
 
 
+def configure_trusted_ca_bundle(ca_cert_path):
+    """
+    Configure cluster-wide trusted CA bundle in Proxy object
+
+    Args:
+        ca_cert_path (str): path to CA Certificate(s) bundle file
```
Comment (vavuthu) [inline ocs_ci/utility/ssl_certs.py:272]: new line after line 272

**Comment by dahorak** [comment]
Fixed review comments and documented newly introduced `EXTERNAL_MODE` parameters.

**Review by vavuthu — APPROVED** [review]

File: ocs_ci/deployment/helpers/external_cluster_helpers.py [hunk 3]
```diff
@@ -187,7 +198,50 @@ def get_rgw_endpoint_api_port(self):
...
+                daemon["ports"][0]
+                for daemon in daemons
+                if daemon["service_name"].startswith("rgw")
+            ][0]
+            # if port doesn't have value, need to check followup way
+            if not port:
+                raise AttributeError(
+                    "Command `cephadm ls` output doesn't have information about rgw port."
+                )
+        except Exception as ex:
```
Comment (ebenahar) [inline ocs_ci/deployment/helpers/external_cluster_helpers.py:212]: Apart from AttributeError, which other exceptions could be caught here?
If it's only AttributeError that we can expect, wouldn't it be simpler to replace the try-except with conditions, such as port doesn't have value and ceph version?

Comment (dahorak) [reply ocs_ci/deployment/helpers/external_cluster_helpers.py:212]: It was mostly copy&paste of the existing code extended to Ceph 6 specifics, but you are right, that it looks quite complicated and that there might be some better way how to structure the code. I'll review it and send update.

Comment (dahorak) [reply ocs_ci/deployment/helpers/external_cluster_helpers.py:212]: I had to get back to the code to remind myself the functionality.
There might be various exceptions raised - basically any step in the `try .. except` block could theoretically fail (`ExternalClusterRGWEndPointMissing` - from `get_rgw_endpoint()`, connection related exceptions from `Connection` object, `KeyError` and `IndexError` and finally the `AttributeError`). In reality, there will be raised only some of those exceptions, which depends on the specifics for each Ceph cluster version (for example in older version the `cephadm` command is not available or the returned structure is different).

File: ocs_ci/deployment/helpers/external_cluster_helpers.py [hunk 4]
```diff
@@ -137,6 +143,25 @@ def upload_exporter_script(self):
...
+    def upload_rgw_cert_ca(self):
+        """
+        Upload RGW Cert CA to RHCS cluster
+
+        Returns:
+            str: absolute path to the CA Cert
+
+        """
+        rgw_cert_ca_path = get_and_apply_rgw_cert_ca()
+        remote_rgw_cert_ca_path = "/tmp/rgw-cert-ca.pem"
```
Comment (petr-balogh) [inline ocs_ci/deployment/helpers/external_cluster_helpers.py:155]: Not sure how this tmp file is used, if it's one time only. If we will need it again, later on, and the machine will reboot, we might loose this file. Is that OK?

Comment (dahorak) [reply ocs_ci/deployment/helpers/external_cluster_helpers.py:155]: It is used one time only - when executed the external cluster exporter script, so it should be ok.

**Pull request merged**